File: roiquery-core/src/main/java/com/roiquery/analytics/utils/ProcessUtils.java
package com.roiquery.analytics.utils;


import android.app.ActivityManager;
import android.app.Application;
import android.content.Context;
import android.text.TextUtils;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.List;


/**
 * <pre>
 *     author: Blankj
 *     blog  : http://blankj.com
 *     time  : 2016/10/18
 *     desc  : utils about process
 * </pre>
 */
public final class ProcessUtils {

    private ProcessUtils() {
        throw new UnsupportedOperationException("u can't instantiate me...");
    }



    /**
     * Return whether app running in the main process.
     *
     * @return {@code true}: yes<br>{@code false}: no
     */
    public static boolean isMainProcess(Application application) {
        return application.getPackageName().equals(getCurrentProcessName(application));
    }

    /**
     * Return the name of current process.
     *
     * @return the name of current process
     */
    public static String getCurrentProcessName(Application application) {
        String name = getCurrentProcessNameByFile();
        if (!TextUtils.isEmpty(name)) return name;
        name = getCurrentProcessNameByAms(application);
        if (!TextUtils.isEmpty(name)) return name;
        name = getCurrentProcessNameByReflect(application);
        return name;
    }

    private static String getCurrentProcessNameByFile() {
        try {
            File file = new File("/proc/" + android.os.Process.myPid() + "/" + "cmdline");
            BufferedReader mBufferedReader = new BufferedReader(new FileReader(file));
            String processName = mBufferedReader.readLine().trim();
            mBufferedReader.close();
            return processName;
        } catch (Exception e) {
            e.printStackTrace();
            return "";
        }
    }

    private static String getCurrentProcessNameByAms(Application application) {
        try {
            ActivityManager am = (ActivityManager) application.getSystemService(Context.ACTIVITY_SERVICE);
            if (am == null) return "";
            List<ActivityManager.RunningAppProcessInfo> info = am.getRunningAppProcesses();
            if (info == null || info.size() == 0) return "";
            int pid = android.os.Process.myPid();
            for (ActivityManager.RunningAppProcessInfo aInfo : info) {
                if (aInfo.pid == pid) {
                    if (aInfo.processName != null) {
                        return aInfo.processName;
                    }
                }
            }
        } catch (Exception e) {
            return "";
        }
        return "";
    }

    private static String getCurrentProcessNameByReflect(Application application) {
        String processName = "";
        try {
            Application app = application;
            Field loadedApkField = app.getClass().getField("mLoadedApk");
            loadedApkField.setAccessible(true);
            Object loadedApk = loadedApkField.get(app);

            Field activityThreadField = loadedApk.getClass().getDeclaredField("mActivityThread");
            activityThreadField.setAccessible(true);
            Object activityThread = activityThreadField.get(loadedApk);

            Method getProcessName = activityThread.getClass().getDeclaredMethod("getProcessName");
            processName = (String) getProcessName.invoke(activityThread);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return processName;
    }
}

File: roiquery-core/src/main/java/com/roiquery/analytics/utils/ProcessUtils.kt
package com.roiquery.analytics.utils

import android.app.Application
import android.content.Context
import android.os.Build
import android.text.TextUtils
import androidx.annotation.RequiresApi

object ProcessUtils {
    fun isInMainProcess(context: Context): Boolean {
        val appContext = context.applicationContext
        val processName = getProcessName(appContext)
        return processName == null || TextUtils.equals(processName, appContext.packageName)
    }

    fun isInServiceProcess(context: Context): Boolean {
        val appContext = context.applicationContext
        val processName = getProcessName(appContext)
        return processName != null && processName.endsWith(":fssvc")
    }

    private var sProcessName: String? = null
    fun getProcessName(context: Context?): String? {
        return if (sProcessName != null) {
            sProcessName
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            processNameAPI28
        } else {
            val applicationContext = context?.applicationContext
            if (applicationContext is Application) {
                sProcessName = getProcessNameViaReflection(applicationContext)
                sProcessName
            } else {
                null
            }
        }
    }

    private val processNameAPI28: String?
         @RequiresApi(Build.VERSION_CODES.P)
         get() = try {
             sProcessName = Application.getProcessName()
             sProcessName
        } catch (var1: Exception) {
            null
        }

    private fun getProcessNameViaReflection(app: Application): String? {
        return try {
            val loadedApkField = app.javaClass.getField("mLoadedApk")
            loadedApkField.isAccessible = true
            val loadedApk = loadedApkField[app]
            val activityThreadField = loadedApk.javaClass.getDeclaredField("mActivityThread")
            activityThreadField.isAccessible = true
            val activityThread = activityThreadField[loadedApk]
            val getProcessName = activityThread.javaClass.getDeclaredMethod("getProcessName", null)
            getProcessName.invoke(activityThread, null) as String
        } catch (var6: Exception) {
            null
        }
    }
}
